File: 2025-10-03 REFATORACAO (Java)/classes/Tributo.java
package JogosVorazes.classes;

public class Tributo {
    private String nome;
    private int saude;
    private Distrito distrito;
    private String segredo;
    private Habilidade[] habilidades;
    private int resistencia;
    private int armadura;

    public Tributo(String nome, int saude, Distrito distrito, String segredo, Habilidade[] habilidades, int resistencia, int armadura) {
        this.nome = nome;
        this.saude = saude;
        this.distrito = distrito;
        this.segredo = segredo;
        this.habilidades = habilidades;
        this.resistencia = resistencia;
        this.armadura = armadura;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public int getSaude() {
        return saude;
    }

    public void setSaude(int saude) {
        this.saude = saude;
    }

    public Distrito getDistrito() {
        return distrito;
    }

    public void setDistrito(Distrito distrito) {
        this.distrito = distrito;
    }

    public String getSegredo() {
        return segredo;
    }

    public void setSegredo(String segredo) {
        this.segredo = segredo;
    }

    public Habilidade[] getHabilidades() {
        return habilidades;
    }

    public void setHabilidades(Habilidade[] habilidades) {
        this.habilidades = habilidades;
    }

    public int getResistencia() {
        return resistencia;
    }

    public void setResistencia(int resistencia) {
        this.resistencia = resistencia;
    }

    public int getArmadura() {
        return armadura;
    }

    public void setArmadura(int armadura) {
        this.armadura = armadura;
    }

    private int calcularDano(int danoRecebido) {
        return danoRecebido-(this.armadura + this.resistencia);
    }

    public void sofrerDano(int danoRecebido) {
        int dano = this.calcularDano(danoRecebido);
        if (dano > 0) {
            if (dano >= this.saude) {
                System.out.println(this.nome + " sofreu um dano fatal! Este personagem está morto. Saúde: " + this.saude + " > " + (this.saude-dano));
            } else {
                System.out.println(this.nome + " sofreu " + dano + " de dano! Saúde: " + this.saude + " > " + (this.saude-dano));
            }
            this.saude = this.saude-dano;
        } else {
            System.out.println("A armadura e resistência de " + this.nome + " foi superior ao dano recebido. Saúde não alterada.");
        }
    }

    public String exibirHabilidades() {
        String[] habilidadesNomes = new String[this.habilidades.length];
        for (int i = 0; i < this.habilidades.length; i++) {
            habilidadesNomes[i] = this.habilidades[i].getNome();
        }
        return String.join(", ", habilidadesNomes);
    }

}

File: 2025-10-03 REFATORACAO/tributo.ts
import { Distrito} from "./distrito";
import { Habilidade } from "./habilidade";

export class Tributo {
    private nome: string;
    private saude: number;
    private distrito: Distrito;
    private segredo: string;
    private habilidades: Habilidade[];
    private resistencia: number;
    private armadura: number;

    constructor(nomeRecebido: string, saudeRecebida: number, distritoRecebido: Distrito, segredoRecebido: string, habilidadesRecebidas: Habilidade[], resistenciaRecebida: number, armaduraRecebida: number) {
        this.nome = nomeRecebido;
        this.saude = saudeRecebida;
        this.distrito = distritoRecebido;
        this.segredo = segredoRecebido;
        this.habilidades = habilidadesRecebidas;
        this.resistencia = resistenciaRecebida
        this.armadura = armaduraRecebida
    }

    public setNome(nomeRecebido: string): void {
        this.nome = nomeRecebido;
    }

    public setSaude(saudeRecebida: number): void {
        this.saude = saudeRecebida;
    }

    public setDistrito(distritoRecebido: Distrito): void {
        this.distrito = distritoRecebido;
    }

    public setSegredo(segredoRecebido: string): void {
        this.segredo = segredoRecebido;
    }

    public setHabilidades(habilidadesRecebidas: Habilidade[]): void {
        this.habilidades = habilidadesRecebidas;
    }

    public getNome(): string {
        return this.nome;
    }

    public getSaude(): number {
        return this.saude;
    }

    public getDistrito(): Distrito {
        return this.distrito;
    }

    public getSegredo(): string {
        return this.segredo;
    }

    public getHabilidades(): void {

        for (let i = 0; i < this.habilidades.length; i++) {
            console.log(this.habilidades[i].getNome())
        }
    }

    public sofrerDano(danoRecebido: number): void {
        const dano = this.calcularDano(danoRecebido)
        if (dano > 0) {
            this.saude = this.saude - dano
            console.log(`${this.nome} sofreu ${dano} de dano! Sua saúde agora é ${this.saude}.`)
        } else {
            console.log(`A armadura e resistência de ${this.nome} foi superior ao dano recebido. Saúde não alterada.`)
        }
    }

    private calcularDano(danoRecebido:number): number {
        const dano:number = danoRecebido-(this.armadura+this.resistencia)
        if (dano > 0) {
            return dano
        } else {
            return 0;
        }
    }



}